### ISSUE TITLE
Problems with unfold_tree

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1146]
`igraph_unfold_tree` seemed unfinished to me. It did not handle disconnected graphs properly when not enough "roots" were passed to it.

Reminder: this function does a BFS traversal of the graph and records the traversal tree. Thus, *if all vertices are visited*, the number of edges of the tree will be the same as the number of edges in the original graph (and the number of vertices one more, as with all trees). The current code actually allocates as edge vector of this size. However, if not all edges are traversed (because a root was not given for all connected components), then some of the elements of the edge vector will not be assigned, and the returned result will be incorrect (or at least weird).

I chose to fix this by allowing the result tree to have isolated nodes (and therefore fewer edges than `vertex_count-1`).

Here's the proposed fix:

https://github.com/szhorvat/igraph/commit/8e840c5cd994f65d67f921ef6e1d23c4220c86ff

A potential problem with this solution: the original code preserved the edges ordering of the original graph. *If* the full graph was traversed, then the `k`th edge of the original graph would correspond to the `k`th edge of the tree. This is no longer true with my patch.

Before sending a PR I wanted to check what the R and Python interfaces do.  Strangely, this happens with both:

```R
> g<-make_ring(4)
> t<-unfold_tree(g,roots=c(1))
> plot(t)
Error in xy.coords(x, y, xlabel, ylabel, log) : 
  'x' is a list, but does not have components 'x' and 'y'
```

```python
g=igraph.Graph.Ring(4)
t=g.unfold_tree(roots=[0])
igraph.plot(t)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/Users/szhorvat/anaconda/envs/net/lib/python2.7/site-packages/IPython/core/formatters.pyc in __call__(self, obj)
    336             method = get_real_method(obj, self.print_method)
    337             if method is not None:
--> 338                 return method()
    339             return None
    340         else:

/Users/szhorvat/anaconda/envs/net/lib/python2.7/site-packages/igraph/drawing/__init__.pyc in _repr_svg_(self)
    347         context = cairo.Context(surface)
    348         # Plot the graph on this context
--> 349         self.redraw(context)
    350         # No idea why this is needed but python crashes without
    351         context.show_page()

/Users/szhorvat/anaconda/envs/net/lib/python2.7/site-packages/igraph/drawing/__init__.pyc in redraw(self, context)
    261             plotter = getattr(obj, "__plot__", None)
    262             if plotter is None:
--> 263                 warn("%s does not support plotting" % obj)
    264             else:
    265                 if opacity < 1.0:

TypeError: not all arguments converted during string formatting
```

I am not really sure about what is going on here, and why the tree is not plottable, even though for this input the result should be correct.

**Comment by szhorvat** [comment]
Maybe the R and Python interfaces use code from before this fix: #1030

**Comment by ntamas** [comment]
I'm not sure what the original purpose of this function was (it was added by @gaborcsardi a long time ago and honestly, I have never used it). It could be the case that the original use-case that he had in mind was to unfold the component(s) containing the root vertices and leave the rest alone. I was always assuming that this function was probably needed by some other algorithm and that's why it was there in the first place, but after scanning the code today for occurrences of `igraph_unfold_tree()`, it looks like this is not the case.

Anyway, I'd rather bend the documentation of the function to the way it works currently than to alter the implementation (unless it acts in a really weird and broken way at the moment).

Regarding the warning that you see in the Python interface: this is unrelated; the problem is that `Graph.unfold_tree()` returns a tuple with two elements (the unfolded graph and a mapping), and this tuple is passed to the formatting operator (`%`) on its own. I will fix it in igraph/python-igraph

**Comment by ntamas** [comment]
Python-related warning fixed in https://github.com/igraph/python-igraph/commit/b570c4c9053cc9f7f7edf58b056fc74739826092

**Comment by szhorvat** [comment]
I assumed that the `unfold_tree` returned a graph without checking carefully. That was my fault. So there is no issue either in R or Python.

**Comment by szhorvat** [comment]
I think the current behaviour does not make much sense because when the graph is not connected (and not enough roots are given), it will effectively insert a bunch of loop-edges onto the first node.

Example in R:

    g<-make_graph(c(1,2, 2,3, 3,1, 4,5), directed = F)
    plot(g)

![image](https://user-images.githubusercontent.com/1212871/50597057-61390e80-0ea6-11e9-8f29-260684c2a9f7.png)

    t<-unfold_tree(g, roots=c(1))$tree
    plot(t)

![image](https://user-images.githubusercontent.com/1212871/50597104-83cb2780-0ea6-11e9-9b98-630e8cba928a.png)

This happens because the function initializes an edge vector with exactly as many elements as the number of edges in the original graph. At initialization, the vector contains all zeros. If not all edges of the graph are traversed during "unfolding", not all entries in the new edge vector get filled. The unfilled part remains all zeros and effectively creates the loop edges.

There is also a comment in the source that it is unfinished.

https://github.com/igraph/igraph/blob/c517ad26e537bbdf363a1f6f9cabd66040664fdd/src/structural_properties.c#L5575

I do agree, however, that there is little use for this function except as a helper for low-level implementation. I exposed it because I thought that a user was asking for something similar (it turned out to be a misunderstanding).

@gaborcsardi : Can you comment on this, so the issue can be closed?

**Comment by szhorvat** [comment]
@gaborcsardi I should have phrased the question differently: What did you have in mind when you added that TODO comment?

**Comment by szhorvat** [comment]
> (@ntamas) I'm not sure what the original purpose of this function was

It turns out that this operation is used sometimes with phylogenetic trees. Here's one example: https://arxiv.org/abs/1506.04438 I recall seeing its use at a conference too (also under the name *unfold*), though I can't google up the papers for that.

Therefore, I was wrong when I said:

> (@szhorvat) I do agree, however, that there is little use for this function except as a helper for low-level implementation.

While I have not found concrete applications for unconnected graphs, it would make sense to have the function behave more reasonable.   The current behaviour is merely an artefact of the specific implementation which never considered disconnected graphs.

> (@ntamas) Anyway, I'd rather bend the documentation of the function to the way it works currently than to alter the implementation (unless it acts in a really weird and broken way at the moment).

Actually, I would say that the current behaviour is really broken.

Let's take another looks at this. I have a patch that simply avoids adding these unnecessary and incorrect self-loops ready, and can send a PR anytime.  The one issue to consider is the edge ordering of the resulting tree (see my original post).

**Comment by szhorvat** [comment]
I'd like to reach a decision on this before the 0.8 release. Our options are:

 1. Leave as it (but the behaviour is broken, this is the worst option)
 2. Use my existing patch and accept that the edge ordering of the tree will no longer correspond to the edge ordering of the input graph
 3. Preserve edge ordering, which requires the whole graph to be traversed. To make this possible, restrict the function to connected graphs. Then we change the function prototype so it takes one root, not several roots.
 4. Preserve the edge ordering. If there isn't a root given for each component, automatically pick a root from that component and use it, so that the full graph can be unfolded. Do return the actually used roots (might require changing the prototype). This will allow the user to determine if any new roots were picked automatically or not.
 5. Split the functionality into `unfold_tree` (connected graph only) and `unfold_forest` (also unconnected)

My personal preference is 4.

The main concern is whose existing use cases we would be breaking.

A found the following uses by Googling:

 - https://buildmedia.readthedocs.org/media/pdf/the-data-driven-ontology-toolkit-ddot/stable/the-data-driven-ontology-toolkit-ddot.pdf
 - ... ?

**Comment by vtraag** [comment]
@szhorvat . I think it would be simplest to restrict to connected graphs only. Providing input for several components separately should not be a problem, and could be left to the user, especially in higher-level languages. I would therefore propose to keep only `unfold_tree`, and restrict it to connected graphs only. Of course, "connected" should mean "reachable" from the provided root(s). If not all vertices were reached, an error should be thrown. Wouldn't that be the simplest approach, and allow to keep the same API?

Actually, now that I am thinking of it, I am not sure at what point it makes sense to provide multiple roots. If I understand the intention of the algorithm correctly, every root that is provided should result in a separate component of the unfolded forest. But what is then the added benefit of providing multiple roots, if you can also simply provide one root, and obtain the root for that node?

Then, thinking along this direction, whether the graph is (weakly) connected or not wouldn't matter, right? If we have the graph `a -> c <- b`, then starting from root `a` we would like to obtain the tree `a -> c` and from `b` we would like to obtain the three `b -> c`. This means that the unfolded tree, starting from `a` does not reach all nodes, but this should not be a problem. Wouldn't that be the simplest? I am not sure how this relates to your proposed fix https://github.com/szhorvat/igraph/commit/8e840c5cd994f65d67f921ef6e1d23c4220c86ff? I am not entirely sure how this relates to the problem of the edge ordering, and why this is a problem?

**Comment by szhorvat** [comment]
>  I think it would be simplest to restrict to connected graphs only. 

> Wouldn't that be the simplest approach, and allow to keep the same API?

The API is designed for multiple roots. It does currently accept multiple roots. It misbehaves only if the entire graph is not reachable from the given roots.

If we were to restrict to "connected" (you're right: reachable from the root) graphs, then the API would need to be changed to not allow more than one root.

>  If I understand the intention of the algorithm correctly, every root that is provided should result in a separate component of the unfolded forest. But what is then the added benefit of providing multiple roots, if you can also simply provide one root, and obtain the tree for that node?

This is what the current implementation sort-of-kind-of does except that it also generates plainly incorrect self-loops (which is just a bug, nothing more). 

My patch simply prevents the self-loops from being generated.

Take this example input graph:

![image](https://user-images.githubusercontent.com/1212871/71886797-1d25bd00-313d-11ea-8cf0-7c52ab45937f.png)

The current code produces this when unfolding, using only `A` as a root:

![image](https://user-images.githubusercontent.com/1212871/71886984-768dec00-313d-11ea-9fe3-21a3217a2f1b.png)

That's fine, except for that spurious self-loop.

**But:** There is a potential issue with my patch. It changes the edge ordering of the generated tree or forest.

Why is the ordering important?

Notice that there is an injective mapping from the edges of the input graph to the edges of the unfolded forest. If the entire graph is traversed, then this is a one-to-one mapping. It may be useful if the `unfold_tree` function provides this mapping in some form.

Currently, it does, through edge ordering. If the entire graph is traversed, then the `k`th edge of the original graph maps to the `k`th edge of the unfolded forest.  If the entire graph is not traversed, then the mapping is not one-to-one (merely injective) therefore it cannot be conveyed by a specific edge ordering of the output.

What the current implementation effectively (and incorrectly) does is that for any edge of the original graph that is not traversed, it *effectively* inserts a self-loop onto the root of the unfolded tree. (It does not do this insertion explicitly. It simply initializes an edge vector to all zeros and then fails to fill in the entire vector, i.e. fails to replace some of those zeros with actual values.)

Thus the only way I could get rid of those self-loops was to also break the edge ordering which retains information about which edge of the forest was induced by which edge of the original tree.

----

Now you may ask:

Is it even important to retain this mapping? I do not know. But it seems reasonable that there are applications out there which rely on this. It may, for example, make it easier to transfer edge attributes from the original graph to the unfolded tree.

I hope this clarified your questions.

**Comment by szhorvat** [comment]
Another, *unrelated* suggestion for `unfold_tree`:

Since the result is a rooted tree / forest, it would make sense to make the result directed even if the input was undirected. The direction of edges encodes the identity of the root.

**Comment by vtraag** [comment]
> If we were to restrict to "connected" (you're right: reachable from the root) graphs, then the API would need to be changed to not allow more than one root.

Yes, my proposal would indeed be to change the API to only accept a single root. Multiple roots essentially corresponds to multiple calls of the function with a single root. I don't see any benefit of having multiple roots in here.

> Is it even important to retain this mapping? I do not know. But it seems reasonable that there are applications out there which rely on this. It may, for example, make it easier to transfer edge attributes from the original graph to the unfolded tree.

Potentially, we could return the `edge_index`, similar to the `vertex_index` used now. That would allow to map which edge has been traversed in the tree. However, at this point I would prefer to simply ignore this and keep it simple (but correct) for 0.8.0. If still desired, we could do so for 0.9.0.

> Since the result is a rooted tree / forest, it would make sense to make the result directed even if the input was undirected. The direction of edges encodes the identity of the root.

I agree, it would make sense to have the edges follow the same direction as the original edge. We could do this for 0.9.0. However, this is not necessary to identify the root, since this can be done using `vertex_index`, right?

**Comment by ntamas** [comment]
My proposal: let's implement something that can be done with nearly-zero effort for 0.8.0, and then fix it properly (according to the proposal of @szhorvat - personally, I prefer option 2 or 4).

I wouldn't be too afraid of breaking things (i.e. not guaranteeing the edge ordering) at this point - there will be other breaking changes as well, so if we can simply get around with plugging in the patch from @szhorvat as a quick remedy and then do option 4 for 0.9.0, that would probably be the best compromise.

**Comment by vtraag** [comment]
Yes, @ntamas, let's simply take the current patch from @szhorvat, and take things further after the release of 0.8.0.

**Comment by ntamas** [comment]
@szhorvat Can you send a PR with the proposed fix so I can merge it?

**Comment by szhorvat** [comment]
I saw the comments, just haven't gotten around to it yet today.

**Comment by ntamas** [comment]
PR is now merged, I think we can reassign this to 0.9.0 for a "proper" fix.
